soci/soci_convert.go

```diff
@@ -0,0 +1,317 @@
+		// Some Registries don't like it when you push a Docker v2 manifest
+		// To increase compatibility, change the media type to OCI image manifest
+		ociIndex.MediaType = ocispec.MediaTypeImageIndex
+		return ociIndex, err
+	}
+	return ocispec.Index{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		MediaType: ocispec.MediaTypeImageIndex,
+		Manifests: []ocispec.Descriptor{
+			img.Target,
+		},
+	}, nil
+}
+
+// annotateImages adds the SOCI index digest to the corresponding image manifest descriptor, modifying `ociIndex` in the process.
+// Adding annotations modifies the image manifest, so annotateImages also pushes the modified image manifests to the blobStore,
+// computes the new image digests, and modified the image descriptors in `ociIndex`
+func (b *IndexBuilder) annotateImages(ctx context.Context, ociIndex *ocispec.Index, sociIndexes []*IndexWithMetadata) error {
+	for i := 0; i < len(ociIndex.Manifests); i++ {
+		manifestDesc := &ociIndex.Manifests[i]
+		// images.Manifest validates the mediatype, no need to do it ourselves like
+		// we did when loading the OCI index
+		manifest, err := images.Manifest(ctx, b.contentStore, *manifestDesc, nil)
+		if err != nil {
+			return err
+		}
+		// Some Registries don't like mixing Docker V2 manifests with OCI image manifests.
+		// Since we use ArtifactTypes for SOCI indexes, we will use OCI image manifests everywhere to increase compatibility.
+		// Registries don't seem to be as picky about layer and config types
+		manifest.MediaType = ocispec.MediaTypeImageManifest
+
+		idx := slices.IndexFunc(sociIndexes, func(i *IndexWithMetadata) bool { return i.ManifestDesc.Digest == manifestDesc.Digest })
+		if idx >= 0 {
+			indexWithMetadata := sociIndexes[idx]
+
+			if manifest.Annotations == nil {
+				manifest.Annotations = make(map[string]string)
+			}
+			manifest.Annotations[ImageAnnotationSociIndexDigest] = indexWithMetadata.Desc.Digest.String()
+		}
+
+		newManifestDesc, err := b.pushOCIObject(ctx, manifest)
+		if err != nil {
+			return err
+		}
+
+		// Modify the original
+		manifestDesc.Digest = newManifestDesc.Digest
+		manifestDesc.Size = newManifestDesc.Size
+		manifestDesc.Annotations = manifest.Annotations
+		manifestDesc.MediaType = ocispec.MediaTypeImageManifest
+
+		if idx >= 0 {
+			indexWithMetadata := sociIndexes[idx]
+			if indexWithMetadata.Desc.Annotations == nil {
+				indexWithMetadata.Desc.Annotations = make(map[string]string)
+			}
+			indexWithMetadata.Desc.Annotations[IndexAnnotationImageManifestDigest] = manifestDesc.Digest.String()
+		}
+	}
+	return nil
+}
+
+// addSociIndexes modifies the list of manifests in the OCI index to include the SOCI indexes.
+// If the OCI index already contains a SOCI index for the same platform, the old SOCI index is replaced,
+// otherwise, the SOCI index is appended to list of manifests.
+func (b *IndexBuilder) addSociIndexes(_ context.Context, ociIndex *ocispec.Index, sociIndexes []*IndexWithMetadata) error {
```

Comment:
Since this is a private function any reason it has to take an unused context var?

Reply:
I was thinking it's easier to take an unused context than to add one later, but you're right that it doesn't matter since this is unexported. 

I think it might make things easier to read in `Convert` if all the steps have the same form, but that's definitely a preference thing.